src/FSharpPlus/Data/NonEmptyList.fs

```diff
@@ -174,24 +300,268 @@ module NonEmptyList =
+    let inline exactlyOne (list: NonEmptyList<'T>) = Seq.exactlyOne list
+
+    /// <summary>Returns a new list with the distinct elements of the input list which do not appear in the itemsToExclude sequence, using generic hash and equality comparisons to compare values.</summary>
+    /// <param name="itemsToExclude">The sequence of items to exclude from the input list.</param>
+    /// <param name="list">The input list.</param>
+    /// <returns>A list that contains the distinct elements of list that do not appear in itemsToExclude.</returns>
+    /// <exception cref="System.ArgumentException">Thrown when itemsToExclude is null.</exception>
+    let inline except (itemsToExclude: #seq<'T>) (list: NonEmptyList<'T>) = Seq.except itemsToExclude list |> ofSeq
+
+    /// <summary>Tests if any element of the list satisfies the given predicate.</summary>
+    /// <param name="predicate">The function to test the input elements.</param>
+    /// <param name="list">The input list.</param>
+    /// <returns>True if any element satisfies the predicate.</returns>
+    let inline exists (predicate: 'T -> bool) (list: NonEmptyList<'T>) = Seq.exists predicate list
+
+    /// <summary>Tests if any pair of corresponding elements of the lists satisfies the given predicate.</summary>
+    /// <param name="predicate">The function to test the input elements.</param>
+    /// <param name="list1">The first input list.</param>
+    /// <param name="list2">The second input list.</param>
+    /// <returns>True if any pair of elements satisfy the predicate.</returns>
+    /// <exception cref="System.ArgumentException">Thrown when the input lists are of different lengths.</exception>    
+    let inline exists2 (predicate: 'T1 -> 'T2 -> bool) (list1: NonEmptyList<'T1>) (list2: NonEmptyList<'T2>) = 
+        Seq.exists2 predicate list1 list2
+
+    /// <summary>Returns a new collection containing only the elements of the collection for which the given predicate returns "true."</summary>
+    /// <param name="predicate">The function to test the input elements.</param>
+    /// <param name="list">The input list.</param>
+    /// <returns>A list containing only the elements that satisfy the predicate.</returns>
+    let inline filter (predicate: 'T -> bool) (list: NonEmptyList<'T>): NonEmptyList<'T> = 
+      list |> toList |> List.filter predicate |> ofList
+
+    /// <summary>Returns a new collection containing only the elements of the collection for which the given predicate returns "true."</summary>
+    /// <param name="predicate">The function to test the input elements.</param>
+    /// <param name="list">The input list.</param>
+    /// <returns>A list containing only the elements that satisfy the predicate.</returns>
+    let inline tryFilter (predicate: 'T -> bool) (list: NonEmptyList<'T>): 'T list = 
+      list |> toList |> List.filter predicate
```

Comment:
Filter-like APIs would need some design discussions.

For example, I would personally prefer `tryFilter` to return `NonEmptyList<'T> option` rather than `'T list`. If I want a `'T list`, I can always write `xs |> NonEmptyList.toList |> List.filter`.

Reply:
I almost did it that way. I thought it made more sense to return `list` because both `list` and `option` have equivalent pattern-matching options (`[]` ~ `None`, `x :: xs` ~ `Some`) with `list` being more flexible and both provide the necessary type safety. However, `list` provided a more sane default for situations like `list |> NonEmptyList.tryFilter fn |> List.sort`. Otherwise, you would be left with something like `list |> NonEmptyList.tryFilter fn |> Option.map NonEmptyList.toList |> Option.defaultValue [] |> List.sort`.

Reply:
What if I changed the names to remove the expectation that an option be returned? Something like `filter` which returns a `list` and `filterUnsafe` which returns a `NonEmptyList`? Or perhaps `filter` and `attemptFilter` or something similar.

Reply:
Again, if you want a sorted `'a list` instead of a sorted `NonEmptyList<'a>`, you can always convert it to `'a list ` first:
```fs
list |> NonEmptyList.toList |> List.filter fn |> List.sort
```

I believe the point of having dedicate `sort/trySort` functions in the `NonEmptyList` is to obtain `NonEmptyList<'a>` as a result.

Note that we do have a generic map function/operator and a generic `monad` computation expression, so if `NonEmptyList.tryFilter` returns `NonEmptyList<'a>`, you can write like these to obtain a filtered and sorted non-empty list:

```fsharp
xs |> NonEmptyList.tryFilter fn |> map NonEmptyList.sort

xs |> NonEmptyList.tryFilter fn |>> NonEmptyList.sort

monad {
  let! ys = xs |> NonEmptyList.tryFilter fn
  return ys |> NonEmptyList.sort
}
```

I strongly believe that `tryFoo` functions should have an `'a option` output, and it's for a good reason: it composes well with the other `tryFoo` functions when used within the computaiton expression:

```fsharp
monad {
  let! xs = tryParse "1,2,3" >>= NonEmptyList.tryOfSeq
  let! ys = trySomeOperation xs
  let! zs = tryAnotherOperation xs
  return! tryYetAnotherOperation ys zs
}
```

As we aim to provide some kind of a standard library, we must consider if the API design provides good ergonomics when used with arbitrary other APIs.

Reply:
Okay. I'll make some updates.